### ISSUE TITLE
Bug: Program hangs / starts dropping audio and stuttering on BT headphones

### ISSUE THREAD

**Issue by rehoehle** [bitgapp/eqMac#946]
**Describe the bug**
So the last few weeks EqMac is very unstable. When using BT headphones the Audio starts to dropout every few seconds. When i disconnect the Headphones it's possible that the Audio is not switching back. Then i happens that the Browser starts hanging when playing Audio. 

So i love the App but in the last times i have to restart my computer much more or stop the app and restart. It's a bit annoying.

**Steps to Reproduce**
Steps to reproduce the behaviour (feel free to change the placeholder as you need):
1.) Connect Bose BT 700 headphones
2.) Put Macbook to sleep and wake up
3.) Or Connect and Disconnect BT headphones and then play a youtube video

**Setup information:**
 - Audio device used for playback: Direct Audio / Bose BT headphones 700
 - Audio transmission interface: Bluetooth or 3.5mm Jack
 - macOS Version: 14.7 and 15
 - eqMac Version: 1.8.7
 - eqMac UI Version: 5.1.4
 - Browser: Brave

After that you have to kill and restart the program to solve the problem thats very annoying. 

Here is an example. When activating the program it starts to hand and is unusable. Same problem in this case with the BT headphones. But it seems that no one cares about problems when paying and i will now stop my subscription. 

https://github.com/user-attachments/assets/f1b87573-0e32-4296-98aa-ee054bbeeb24

**Comment by rehoehle** [comment]
Actually Netflix is stuttering and eqMac has load of 3000. Thats all really annoying and not really usable. 

![2024-09-19_20-56](https://github.com/user-attachments/assets/0b442e20-11ee-4671-9bc5-7071109fd51d)

**Comment by rehoehle** [comment]
I have the same problem with the newest Version 1.8.8 this was a question in another ticket. Actually when i try to play a Youtube video it's stuttering. Other Applications working well. It seems that the is a problem with the Chrome Browser. When i start a video it's stuttering and the load goes to 3000. Mainly after the standby. Now i have to restart. 

Same with BT it's stuttering every few seconds.

**Comment by ragauskl** [comment]
We're still looking into this and similar issues. Rest assured it's not just ignored, but the biggest issue is that we can't replicate this at all on 3 mac devices with different macOS versions and with multiple headphones making it much more difficult to debug and fix. 

Currently I can't say if it's issue with bitrates, dev may know more about this but currently he's occupied with the update and I'd like to not interrupt his focus, but I'll ask when he'll have time it (I usually accumulate list of questions before bothering him).  

The update he's working on however does include a big change to the driver and we're hoping this may also resolve stuttering and stop/play issues some people are experiencing. 

It's not the case of 'no one cares about problems', we're just doing what's possible with current resources which are very limited. In meantime, I definitely understand your decision to cancel the subscription, it is an annoying issue to deal with.

**Comment by rehoehle** [comment]
@ragauskl if you need some logs or something else then tell me. The problem is when no one reacts on the tickets here in a week then no one knows that you working on it.

**Comment by ragauskl** [comment]
Will do. Yeah - github does tend to fall back in priority. Roman is mostly busy with the update, so I'm helping for now but I'm also busy with my own project and customer support, Crisp comes as priority and has most users to reply to, so Github and Discord remain a last priority if there's time unless someone directly links from Crisp to an issue, as both are 'community' support channels. We had a hired customer support for short time, but that's gone for now/didn't work out long term.

**Comment by kersten-lohmeyer** [comment]
I only had this stuttering since upgrading to macOS Sequoia. No problems before.

**Comment by ragauskl** [comment]
@kersten-lohmeyer Are you on eqMac 1.8.7? If yes try enabling Beta Program and update to 1.8.8 (or wait for it to be released in stable, should be any day now I think)

**Comment by kersten-lohmeyer** [comment]
> @kersten-lohmeyer Are you on eqMac 1.8.7? If yes try enabling Beta Program and update to 1.8.8 (or wait for it to be released in stable, should be any day now I think)

Yes I am. I wait for stable, because it doesn't happen too often. Thanks!

**Comment by rehoehle** [comment]
@kersten-lohmeyer i have installed the 1.8.8 it has the same problem for me. So actually the app is not really working. Something is broken in the App that causes a high load and hanging of the app.

**Comment by ragauskl** [comment]
Some changes to driver are pending to be released with 1.8.9 (first in beta), there's some hope it may help with similar issues as it's partial rewrite/change of how it works. I can't promise anything tho as we can't replicate the issue in first place making it harder to fix it

**Comment by kersten-lohmeyer** [comment]
> Some changes to driver are pending to be released with 1.8.9 (first in beta), there's some hope it may help with similar issues as it's partial rewrite/change of how it works. I can't promise anything tho as we can't replicate the issue in first place making it harder to fix it

What changed so much in Sequoia that a rewrite is needed?

**Comment by ragauskl** [comment]
It's not a full on rewrite, more of restructuring/moving things around. Previous version could have also benefited from this but only with Sequoia the issue became too consistent/obvious

**Comment by ragauskl** [comment]
Btw, with youtube and audio stuttering/dropping out - I also noticed this  just now when using BT headphones with Windows and watching youtube stream (already preloaded so not internet problem). I tried switching to different BT headphones and random stuttering was still happening, only when I used built in speakers the stuttering is gone. 

Maybe on mac there's something to do with eqMac at least for some but I think there could be something else as well, as I definetly don't have eqMac on Windows, no EQ or other audio app either. Just browser and youtube open.

**Comment by kersten-lohmeyer** [comment]
I have the stuttering with a USB-C audio interface (Motu M4). No BT-Headphones.

**Comment by rehoehle** [comment]
@ragauskl thats a BT problem. I have eqMac running with the normal Audio Jack and it's working. When i now enable BT it's stuttering and broken. When i disable eqMac and use BT it's working.

**Comment by ragauskl** [comment]
@rehoehle fair enough, just wanted to ask to disable eqMac and see if it still stutters but you checked that.

In my case I think it's still something to do with youtube/browser because I use the BT daily for music from other platforms and never noticed an issue, so it's not only BT problem. On other device however which I use more often for youtube - that's not an issue, a Windows device as well.

**Comment by rehoehle** [comment]
@ragauskl now i have checked with my new Airpods Pro 2. But same problem. So it seems that the app run into an issue and hang. When do you plan to bring the update?

**Comment by ragauskl** [comment]
@rehoehle I can't tell you for sure unfortunately. nodeful is currently working on the driver changes which hopefully should resolve this, but currently he's dealing with a major issue of macOS not behaving in the way it's promising in official documentation causing tiny, but very noticeable in audio, timing issues. He already tried many ways to fix this with no luck and is frustrated by the issue just as much if not more as users dealing with existing issues. He'll figure it out but as he's the sole person behind eqMac and only one working on this it can take some time and I can't say how much. I only help with customer support when I can so he can focus on the development

**Comment by ragauskl** [comment]
Looks like nodeful resolved the mentioned issue, there's are some more issues he wants to resolve before an update in beta channel is released. This does not guarantee fix for the issue as again we couldn't replicate it under same circumstances like other users, but we certainly hope this will help.

**Comment by rehoehle** [comment]
@ragauskl Thanks i can then check it when the version is available and tell you if it's working. If he needs more informations he can contact me.

**Comment by rehoehle** [comment]
@ragauskl and @nodeful i think something changed with the new MacOS 18.0.1 update. The change to BT seems to work now. So it takes a short time to change but the App is not going to get a load of 2000. If think with your new update this could be fixing the general problem.

**Comment by rehoehle** [comment]
@ragauskl ok interesting it's working after an restart but only until the standby on the next day it's not working and same problem.

**Comment by rehoehle** [comment]
Interesting problem this problem exists only after Standby. After restart to the next Standby it's working well.

**Comment by ragauskl** [comment]
Could be that because after restart the driver starts from fresh state, after standby however only the UI can be reloaded and reloading driver is not as seamless so it's not done automatically to reset the state. After standby the driver is doing additional stuff to reset state where possible without reload, during which time slightest mistakes like race conditions start to become much more apparent - this is also very device environment dependant, and on some devices may not occur at all or very rarely making it harder to debug.

**Comment by kersten-lohmeyer** [comment]
I switched to beta and it was working fine for a very long time but now it happened again. I had a Youtube video open in Chrome and I switched from my Motu M4 audio interface to MacBook Pro speakers and the video stuttered. But now, after writing this post, I switched back to Chrome, resumed playback and everything was fine again.

**Comment by rootshellz** [comment]
I am also experiencing the stuttering audio issue on a USB audio interface (Focusrite Scarlett 4i4).  I encounter this on app v1.8.8 w/ driver v2.5.4 on Sequoia (15.0.1).  For what it's worth, I have not yet seen the issue with the same setup on a Sonoma (14.7) machine.

**Comment by devtcp** [comment]
from my observation, the stuttering is easier to be reproduced by having some google chrome playing some youtube video in the background (listening to music), pausing it and then instantly lock the laptop, and wait for the laptop to go to sleep on its own.(i believe that causes the issues but i might be wrong)

the only viable solution so far for me without rebooting the laptop is 

- go to activity monitor, under CPU processes, look for coreaudio
- then force quit coreaudiod
- click on eqmac and on the bottom of the selected audio interface would be empty. 
- click on the "refresh" button on the bottom left, then your previously selected interface should appear automatically
- your audio should be working again without stutter.

**Comment by rehoehle** [comment]
@ragauskl when do you think is here a new version available?

**Comment by ragauskl** [comment]
@devtcp That likely helps because it restarts the driver, you can also try the steps bellow (or just the commands) if it's easier/faster to do

1. Stop all audio playback
2. Quit eqmac
4. Run in terminal
   - For soft reset: "sudo launchctl kickstart -k system/com.apple.audio.coreaudiod"
   - Or for hard reset: "sudo killall coreaudiod"
   
@rehoehle The core work has been complete, dev is polishing off about 4 small bugs right now. While they are not as major as audio stuttering dev does not want to release update until everything is working at least for him. We had a more in-depth conversation regarding this and one of current problems why he doesn't want to release a beta right now until everything is polished out (which constantly keeps increasing the time needed, as he fixes one thing but then finds another issue - the main reason for this is also the fact the there has been a major change in driver which has made a list of previous issues that were occurring to few users much more obvious/consistent) is that there is already large amount of people in beta so he's afraid that while it will fix issue like stuttering it may break something for many other users. So the 'beta' is now more of 'release candidate', I suggested making a build for at least one person to try and confirm if this issue specifically has been resolved, but he disagreed, however we discussed that beta/stable release should be changed to allow users to also opt in to 'alpha' release and alpha may be closed/invite only. That would allow to release the update as is to check if changes fixed any one specific issue even if other things are still broken. 

So alpha will be added after the update, but for now the update itself remains a priority to complete. As for when it will be completed, due to everything explained above I can't provide time estimate still, last 4 weeks it has been 'any week now' and a new issue pops up, these are also a lot of times issues that don't have a definitive answer or obvious solution and require a lot of debugging and sorting through a lot of information the app is outputting and can take less than a day or multiple days per each issue. 

I wish I could provide a better estimate, but unfortunately that is not possible at the moment, I can only say that dev is working every day to resolve this as soon as possible.

**Comment by rehoehle** [comment]
@ragauskl i hope i can test it. At the moment it's unusable. So everything is better then this.

**Comment by ragauskl** [comment]
I understand, I tried to communicate this to the dev but last word is his

**Comment by 1chappie** [comment]
Experiencing the same issue on Sequoia 15.0.1 on every output device
eqMac Version: 1.8.8, Driver 2.5.4

Sometimes it fixes itself after I switch the output device while audio is playing and eqmac is on, other times it fixes itself after turning eqmac off and on again, another time after quitting eqmac and reopening the application, another time after using "Restart eqMac", and 2 times I had to log out and log back into macos to fix it. Sorry if this isn't very specific, but I can't seem to reproduce an exact way to make it work, it's all trial and error. 
`sudo killall coreaudiod` seems to work consistently.

**Comment by ragauskl** [comment]
1.8.9 now available in beta, please do keep in mind that it's beta and it includes major underlaying changes to the driver, developer fixed any issues he could replicate (note that this issue could not be replicate under same circumstances, therefore it's 50/50 if the update will help you or not), but that does not mean that new issues will not occur to others or that all issues are definitely fixed. Any feedback is valuable so that we can know if it's working as expected to everyone and what issues are still present.

**Comment by rehoehle** [comment]
@ragauskl at the moment it seems good. Until now it's working and now problems.

**Comment by ragauskl** [comment]
@rehoehle Nice, fingers crossed it will stay that way

**Comment by ragauskl** [comment]
@rehoehle 1.8.10 is out now - see if that helps at all

**Comment by rehoehle** [comment]
@ragauskl thank i have installed it now and will check it now :)

**Comment by boriselec** [comment]
Came here with same problem, bt headphones stuttering.
Reinstalling app/driver helped.
app 1.8.8
driver 2.5.4

**Comment by ragauskl** [comment]
@boriselec try 1.8.10 by enabling 'beta program' in settings and checking for updates

**Comment by 4S1ght** [comment]
I've got the same stuttering issues, except in my case they appear every single time I change the output device, be it bluetooth, built-in speakers, audio jack, whatever - Any change does it. It's worth noting that I have automatic profile switching enabled, so each time I switch the output device a different equalizer profile is loaded. This has been happening for about the last half a year...

**Comment by ragauskl** [comment]
@4S1ght  what eqMac version are you using?

**Comment by ragauskl** [comment]
That's because it was only fixed in 1.8.10, we were not able to replicate (therefore fix) the problem before, what changed is that we bought a DAC with 3+ channels and could replicate the issue finally

**Comment by divStar** [comment]
I have the AirPods Max (not sure about the model number, but it's USB-C already) and it works flawlessly so far. I'm on App version `1.8.8` and driver `2.5.4` and no stuttering so far. I was a bit worried about it draining the battery more, but it seems to not be the case (at least for now).

**Comment by ragauskl** [comment]
Closing as complete as multiple users including creator of the issue have confirmed that it's working in 1.8.11. If issue reoccurs feel free to open new issue or reopen this one

**Comment by marktoman** [comment]
This is still happening. You guys can't reproduce it? I've the problem with all 3 headphones, each from a different maker.

**Comment by rehoehle** [comment]
@marktoman do you have the lastest beta? So for me the problem is solved. I have now some other problems.

**Comment by marktoman** [comment]
@rehoehle Ah, it doesn't occur in beta.

**Comment by franko-franicevich** [comment]
Just as another data point, I had issues when connecting my bluetooth headset, updating to the 1.8.11 beta seems to have resolved it. Switching back and forth without an issue now.

**Comment by mattlacey** [comment]
Beta just solved the same problem for me using Display Port audio.
